File: sdk/nodejs/src/util/base-annotations.ts
// Constants

import {Type} from "../model/resource";

export const Enabled = "true";

// Resource flags
export const KeepHistory = "KeepHistory";
export const HistoryResource = "HistoryResource";
export const NormalizedResource = "NormalizedResource";
export const AutoCreated = "AutoCreated";
export const EnableAudit = "EnableAudit";
export const DisableVersion = "DisableVersion";
export const DisableBackup = "DisableBackup";

// Property params
export const SourceDef = "SourceDef";
export const SourceIdentity = "SourceIdentity";
export const SourceMatchKey = "SourceMatchKey";
export const Identity = "Identity";
export const SpecialProperty = "SpecialProperty";

// HCL
export const IsHclLabel = "IsHclLabel";
export const HclBlock = "HclBlock";

// Request annotations
export const IgnoreIfExists = "IgnoreIfExists";
export const CheckVersion = "CheckVersion";

// Security
export const AllowPublicAccess = "AllowPublicAccess";
export const AllowPublicReadAccess = "AllowPublicGetAccess";
export const AllowPublicCreateAccess = "AllowPublicCreateAccess";
export const AllowPublicUpdateAccess = "AllowPublicUpdateAccess";
export const AllowPublicDeleteAccess = "AllowPublicDeleteAccess";

// SQL
export const SQLType = "SQLType";
export const SQLUseTextType = "SQLUseTextType";
export const UseJoinTable = "UseJoinTable";

// Code generator
export const TypeName = "TypeName";
export const SelfContainedProperty = "SelfContainedProperty";
export const AllowEmptyPrimitive = "AllowEmptyPrimitive";
export const CommonType = "CommonType";

// REST API
export const RestApiDisabled = "RestApiDisabled";

// OpenAPI
export const OpenApiGroup = "OpenApiGroup";
export const OpenApiHide = "OpenApiHide";
export const OpenApiRestPath = "OpenApiRestPath";

// Service
export const ServiceKey = "ServiceKey";

// ExtensionId
export const ExtensionId = "ExtensionId";

// Bypass extensions
export const BypassExtensions = "BypassExtensions";

export const StudioSeparatePages = "StudioSeparatePages";

// Client allowed annotations
export const ClientAllowedAnnotations = {
    [BypassExtensions]: true,
};

export interface AnnotationDef {
    name: string
    type: Type
    defaultValue?: any
}

const annotations: AnnotationDef[] = [
    {name: Enabled, type: Type.BOOL},
    {name: KeepHistory, type: Type.BOOL},
    {name: HistoryResource, type: Type.BOOL},
    {name: NormalizedResource, type: Type.BOOL},
    {name: AutoCreated, type: Type.BOOL},
    {name: EnableAudit, type: Type.BOOL},
    {name: DisableVersion, type: Type.BOOL},
    {name: DisableBackup, type: Type.BOOL},
    {name: SourceDef, type: Type.BOOL},
    {name: SourceIdentity, type: Type.BOOL},
    {name: SourceMatchKey, type: Type.BOOL},
    {name: Identity, type: Type.BOOL},
    {name: SpecialProperty, type: Type.BOOL},
    {name: IsHclLabel, type: Type.BOOL},
    {name: HclBlock, type: Type.BOOL},
    {name: IgnoreIfExists, type: Type.BOOL},
    {name: CheckVersion, type: Type.BOOL},
    {name: AllowPublicAccess, type: Type.BOOL},
    {name: AllowPublicReadAccess, type: Type.BOOL},
    {name: AllowPublicCreateAccess, type: Type.BOOL},
    {name: AllowPublicUpdateAccess, type: Type.BOOL},
    {name: AllowPublicDeleteAccess, type: Type.BOOL},
    {name: SQLType, type: Type.BOOL},
    {name: SQLUseTextType, type: Type.BOOL},
    {name: UseJoinTable, type: Type.BOOL},
    {name: TypeName, type: Type.BOOL},
    {name: SelfContainedProperty, type: Type.BOOL},
    {name: AllowEmptyPrimitive, type: Type.BOOL},
    {name: CommonType, type: Type.BOOL},
    {name: RestApiDisabled, type: Type.BOOL},
    {name: OpenApiGroup, type: Type.BOOL},
    {name: OpenApiHide, type: Type.BOOL},
    {name: OpenApiRestPath, type: Type.BOOL},
    {name: ServiceKey, type: Type.BOOL},
    {name: ExtensionId, type: Type.BOOL},
    {name: BypassExtensions, type: Type.BOOL},
    {name: StudioSeparatePages, type: Type.BOOL}
];

File: pkg/service/annotations/base-annotations.go
package annotations

// values:
// true
const Enabled = "true"

// resource flags
const KeepHistory = "KeepHistory"
const HistoryResource = "HistoryResource"
const AutoCreated = "AutoCreated"
const EnableAudit = "EnableAudit"
const DisableVersion = "DisableVersion"
const DisableBackup = "DisableBackup"

// property params
const SourceDef = "SourceDef"
const SourceIdentity = "SourceIdentity"
const SourceMatchKey = "SourceMatchKey"
const Identity = "Identity"
const SpecialProperty = "SpecialProperty"

// hcl
const IsHclLabel = "IsHclLabel"
const HclBlock = "HclBlock"

// request annotations
const IgnoreIfExists = "IgnoreIfExists"
const CheckVersion = "CheckVersion"

// security
const AllowPublicAccess = "AllowPublicAccess"
const AllowPublicReadAccess = "AllowPublicGetAccess"
const AllowPublicCreateAccess = "AllowPublicCreateAccess"
const AllowPublicUpdateAccess = "AllowPublicUpdateAccess"
const AllowPublicDeleteAccess = "AllowPublicDeleteAccess"

// sql
const SQLType = "SQLType"
const SQLUseTextType = "SQLUseTextType"
const UseJoinTable = "UseJoinTable"

// code generator
const TypeName = "TypeName"
const SelfContainedProperty = "SelfContainedProperty"
const AllowEmptyPrimitive = "AllowEmptyPrimitive"
const CommonType = "CommonType"

// restapi
const RestApiDisabled = "RestApiDisabled"

// openapi
const OpenApiGroup = "OpenApiGroup"
const OpenApiHide = "OpenApiHide"
const OpenApiRestPath = "OpenApiRestPath"

// service
const ServiceKey = "ServiceKey"

// extensionId
const ExtensionId = "ExtensionId"

// bypass extensions
const BypassExtensions = "BypassExtensions"

var ClientAllowedAnnotations = map[string]bool{
	BypassExtensions: true,
}
